### PR TITLE
feefrac: 128-bit multiply support in MSVC

### PR DESCRIPTION
**Pull request by sipa**
Feerate comparisons in the recently (#29242) introduced `FeeFrac` type rely on multiplications between 32-bit and 64-bit integers. On 64-bit systems, hardware can do this natively. On GCC and Clang we can use the `__int128` type for this, but on MSVC one needs to use the `_mul128` or `_mulh` intrinsics instead. This PR adds the use of `_mul128` which is available on x86_64 systems.

Performance of these operations isn't currently very important, but they will become crucial with cluster mempool.

I have not tested this code myself, though it's based on similar code in libsecp256k1 (see https://github.com/bitcoin-core/secp256k1/blob/v0.4.1/src/int128_struct_impl.h#L7L30).

### CODE REVIEW

**Review by hebasto — COMMENTED** [review]
Concept ACK.

File: src/util/feefrac.h [hunk 1]
```diff
@@ -11,6 +11,10 @@
 #include <span.h>
 #include <util/check.h>
 
+#if defined(_MSC_VER) && (defined(_M_X64) || defined(_M_ARM64))
```
[the remark this answers is not in the document]
Comment (hebasto) [reply src/util/feefrac.h:14]: Yes, we have no tests for MSVC builds on ARM64.

**Comment by dergoegge** [comment]
> Hmm, does our CI not run the fuzz corpus when building with MSVC?

Afaik it does not but it would probably make sense. We also do it in the MacOS job.

**Comment by hebasto** [comment]
> Hmm, does our CI not run the fuzz corpus when building with MSVC?

No, it does not. Even a fuzz binary is not built.

File: src/util/feefrac.h [hunk 2]
```diff
@@ -53,10 +57,19 @@ struct FeeFrac
...
-        // If __int128 is available, use 128-bit wide multiply.
+        // If __int128 is available, use 128-bit wide multiplication.
         return __int128{a} * b;
     }
+#elif defined(_MSC_VER) && defined(_M_X64)
+    static inline std::pair<int64_t, uint64_t> Mul(int64_t a, int32_t b) noexcept
+    {
+        // On 64-bit MSVC, use _mul128 intrinsic for wide multiplication.
+        std::pair<int64_t, uint64_t> ret;
+        ret.second = _mul128(a, b, &ret.first);
```
Comment (theuni) [inline src/util/feefrac.h:68]: To help other reviewers:
```c++
__int64 _mul128(
   __int64 Multiplier,
   __int64 Multiplicand,
   __int64 *HighProduct
);

Return value
The low 64 bits of the product.
```

So this should result in:
ret = <high, low>

Which matches the other impls.

**Review by theuni — APPROVED** [review]
utACK 5fb70b5b72c96a7e640900de56393b0bdb9df37a

(I didn't look into the additional unit tests)

Godbolt output here:
https://godbolt.org/z/8Msz984bv

Compiles down to roughly the same.

File: src/util/feefrac.h [hunk 3]
```diff
@@ -53,10 +57,19 @@ struct FeeFrac
 #ifdef __SIZEOF_INT128__
     static inline __int128 Mul(int64_t a, int32_t b) noexcept
     {
-        // If __int128 is available, use 128-bit wide multiply.
+        // If __int128 is available, use 128-bit wide multiplication.
         return __int128{a} * b;
     }
+#elif defined(_MSC_VER) && defined(_M_X64)
+    static inline std::pair<int64_t, uint64_t> Mul(int64_t a, int32_t b) noexcept
```
Comment (hebasto) [inline src/util/feefrac.h:64]: Why do this `Mul`'s return type and `MulFallback`'s one differ?

**Comment by hebasto** [comment]
I've add a [benchmark](https://github.com/hebasto/bitcoin/commit/0f5abaf980122cce1bd0d405d17db3bfab273afa) for the `FeeRateCompare` function, which calls `Mul` twice.

Here are results on Windows 11, `Release` configuration, which implies `/O2 /Oi` compile flags:
- master branch @ 61de64df6790077857faba84796bb874b59c5d15:
```
> .\build_msvc\master\bench_bitcoin.exe -filter=FeefracMultipication

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                1.47 |      680,300,487.93 |    1.9% |      0.17 | `FeefracMultipication`
> .\build_msvc\master\bench_bitcoin.exe -filter=FeefracMultipication

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                1.47 |      679,903,607.75 |    1.1% |      0.18 | `FeefracMultipication`
> .\build_msvc\master\bench_bitcoin.exe -filter=FeefracMultipication

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                1.43 |      698,557,147.13 |    1.3% |      0.17 | `FeefracMultipication`
```
- this PR:
```
> .\build_msvc\pr29758\bench_bitcoin.exe -filter=FeefracMultipication

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                1.53 |      654,804,351.62 |    0.7% |      0.18 | `FeefracMultipication`
> .\build_msvc\pr29758\bench_bitcoin.exe -filter=FeefracMultipication

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                1.52 |      659,374,380.82 |    0.7% |      0.18 | `FeefracMultipication`
> .\build_msvc\pr29758\bench_bitcoin.exe -filter=FeefracMultipication

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                1.51 |      660,797,183.37 |    0.6% |      0.18 | `FeefracMultipication`
```

The performance worsened by ~4%.

cc @sipsorcery 

---

FWIW, this benchmark shows <10% of performance improvement for `__int128` implementation comparing to the naive one on Linux.

File: src/util/feefrac.h [hunk 4]
```diff
@@ -53,10 +57,19 @@ struct FeeFrac
...
+    {
+        // On 64-bit MSVC, use _mul128 intrinsic for wide multiplication.
+        std::pair<int64_t, uint64_t> ret;
+        ret.second = _mul128(a, b, &ret.first);
+        return ret;
```
Comment (hebasto) [inline src/util/feefrac.h:69]: > on MSVC one needs to use the `_mul128` or `_mulh` intrinsics instead

From my research, it follows that [`__mulh`](https://learn.microsoft.com/en-us/cpp/intrinsics/mulh) seems more performant.

This code:
```suggestion
        // On 64-bit MSVC, use __mulh intrinsic for wide multiplication.
        return {__mulh(a, b), std::bit_cast<uint64_t>(a) * b};
```
gives the following numbers for the [benchmark](https://github.com/bitcoin/bitcoin/pull/29758#issuecomment-2147006298):
```
> build_msvc\x64\Release\bench_bitcoin.exe -filter=FeefracMultipication

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                1.41 |      707,119,862.93 |    2.6% |      0.17 | `FeefracMultipication`
> build_msvc\x64\Release\bench_bitcoin.exe -filter=FeefracMultipication

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                1.36 |      734,209,621.15 |    0.9% |      0.16 | `FeefracMultipication`
> build_msvc\x64\Release\bench_bitcoin.exe -filter=FeefracMultipication

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                1.39 |      720,775,228.66 |    3.1% |      0.17 | `FeefracMultipication`
```

**Comment by hebasto** [comment]
@sipa
> Would you mind benchmarking with smaller realistic numbers (say fee and size both between 0 and 2^30)?

I switched to `FastRandomContext::randbits(30)`.

> If the top limb is equal, it's possible the naive code does worse.

Numbers have not changed:
- the master branch:
```
> build_msvc\x64\Release\bench_bitcoin.exe -filter=FeefracMultipication

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                1.47 |      682,496,667.82 |    1.1% |      0.17 | `FeefracMultipication`
> build_msvc\x64\Release\bench_bitcoin.exe -filter=FeefracMultipication

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                1.46 |      686,538,677.70 |    1.3% |      0.17 | `FeefracMultipication`
> build_msvc\x64\Release\bench_bitcoin.exe -filter=FeefracMultipication

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                1.43 |      701,448,254.82 |    1.5% |      0.17 | `FeefracMultipication`
```
- this PR:
```
> build_msvc\x64\Release\bench_bitcoin.exe -filter=FeefracMultipication

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                1.55 |      646,069,682.69 |    0.8% |      0.18 | `FeefracMultipication`
> build_msvc\x64\Release\bench_bitcoin.exe -filter=FeefracMultipication

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                1.52 |      656,962,444.61 |    0.6% |      0.18 | `FeefracMultipication`
> build_msvc\x64\Release\bench_bitcoin.exe -filter=FeefracMultipication

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                1.53 |      653,076,477.59 |    0.9% |      0.18 | `FeefracMultipication`
```

**Pull request closed without merging**